crates/engine_bibtex/bibtex/bibtex.c

```diff
@@ -5480,1554 +3615,1496 @@ static void get_aux_command_and_process(void)
+                set_old_num_cites(num_cites());
+                if (ctx->all_entries) {
+                    set_cite_ptr(all_marker());
+                    while (cite_ptr() < old_num_cites()) {
+                        set_cite_info(cite_ptr(), cite_list(cite_ptr()));
+                        set_entry_exists(cite_ptr(), false);
+                        set_cite_ptr(cite_ptr() + 1);
                     }
-                    cite_ptr = all_marker;
+                    set_cite_ptr(all_marker());
                 } else {
 
-                    cite_ptr = num_cites;
-                    all_marker = 0 /*any_value */ ;
+                    set_cite_ptr(num_cites());
+                    set_all_marker(0 /*any_value */);
                 }
             }
         }
-        read_performed = true;
-        bib_ptr = 0;
-        while (bib_ptr < num_bib_files) {
-
-            if (verbose) {
-                printf_log("Database file #%ld: ", (long) bib_ptr + 1);
-                print_bib_name();
+        ctx->read_performed = true;
+        set_bib_ptr(0);
+        while (bib_ptr() < ctx->num_bib_files) {
+            if (ctx->config.verbose) {
+                printf_log("Database file #%ld: ", (long) bib_ptr() + 1);
+                TRY(print_bib_name());
             } else {
                 char buf[512];
-                snprintf(buf, sizeof(buf) - 1, "Database file #%ld: ", (long) bib_ptr + 1);
-                ttstub_output_write (log_file, buf, strlen(buf));
-                log_pr_bib_name();
-            }
-            bib_line_num = 0;
-            buf_ptr2 = last;
-            while (!eof(bib_file[bib_ptr]))
-                get_bib_command_or_entry_and_process();
-            peekable_close(bib_file[bib_ptr]);
-            bib_file[bib_ptr] = NULL;
-            bib_ptr = bib_ptr + 1;
-        }
-        reading_completed = true;
-        ;
+                snprintf(buf, sizeof(buf) - 1, "Database file #%ld: ", (long) bib_ptr() + 1);
+                bib_log_prints(buf);
+                TRY(log_pr_bib_name());
+            }
+            set_bib_line_num(0);
+            bib_set_buf_offset(BUF_TY_BASE, 2, bib_buf_len(BUF_TY_BASE));
+            hash_loc cur_macro_loc = 0;
+            hash_loc field_name_loc = 0;
+            while (!eof(cur_bib_file()))
+                get_bib_command_or_entry_and_process(ctx, &cur_macro_loc, &field_name_loc);
+            peekable_close(cur_bib_file());
+            set_cur_bib_file(NULL);
+            set_bib_ptr(bib_ptr() + 1);
+        }
+        ctx->reading_completed = true;
 
         {
-            num_cites = cite_ptr;
-            num_preamble_strings = preamble_ptr;
+            set_num_cites(cite_ptr());
+            ctx->num_preamble_strings = preamble_ptr();
             {
-                if ((num_cites - 1) * num_fields + crossref_num >= max_fields) {
+                CiteNumber cites = num_cites();
+                if (cites > 0) {
+                    cites -= 1;
```

Comment:
This bug is like, a micro-example of two of the big pros of Rust here

A) this would have caused a clear panic instead of silent underflow on debug mode, making the problem immediately obvious
B) There's a dedicated method on integers, `saturating_sub`, which does this but shows the intent a lot better

Reply:
Word! Once you get used to the stricter behavior of a language like Rust, I feel like it becomes more and more amazing that we all put up with C's anything-goes style for so long.